### PR TITLE
Refactor kubelet syncPod method and fix behavior of Infra Containers for RestartPolicy::Never.

### PR DESCRIPTION
**Pull request by gmarek**
WIP version of refactoring needed by #4731. Tests obviously don't pass, as this PR changes order and number of methods called. I'll fix them when we agree on the finial version of the flow. I'll be grateful if someone who knows more what's going on on the node will take a look, and see if this change even remotely makes sense. I tried to keep old semantics of things that I don't understand, but I changed some in places I thought I do.

### CODE REVIEW

**Comment by erictune** [comment]
@yujuhong do you want to take a first look at this?  Feel free to reassign if you need to.

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1003,18 +1003,7 @@ func (kl *Kubelet) killContainersInPod(pod *api.BoundPod, dockerContainers docke
...
-	containersToKeep := make(map[dockertools.DockerID]empty)
-	killedContainers := make(map[dockertools.DockerID]empty)
-	glog.V(4).Infof("Syncing Pod, podFullName: %q, uid: %q", podFullName, uid)
-
-	ref, err := api.GetReference(pod)
-	if err != nil {
-		glog.Errorf("Couldn't make a ref to pod %q: '%v'", podFullName, err)
-	}
-
+func (kl *Kubelet) MakeDataDirs(uid types.UID) error {
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: this probably doesn't have to be public

**Comment by vmarmol** [comment]
One overall comment I would add is that it may be worthwhile to make some of these changes separate and smaller PRs. This code is being changed by ~3 people concurrently and we've had enough trouble with changes to it that I err on the side of caution :)

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1025,137 +1014,204 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
...
-		glog.V(2).Infof("Pod infra container doesn't exist for pod %q, killing and re-creating the pod", podFullName)
-		count, err := kl.killContainersInPod(pod, dockerContainers)
-		if err != nil {
-			return err
-		}
-		podInfraContainerID, err = kl.createPodInfraContainer(pod)
-		if err != nil {
-			glog.Errorf("Failed to introspect pod infra container: %v; Skipping pod %q", err, podFullName)
-			return err
+func (kl *Kubelet) ShouldContainerBeRestarted(pod *api.BoundPod, containerName, dockerContainerName string, uid types.UID) bool {
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Same here about being private

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -1510,7 +1547,7 @@ func (kl *Kubelet) updatePods(u PodUpdate, podSyncTypes map[types.UID]metrics.Sy
 		kl.pods = u.Pods
 		kl.pods = filterHostPortConflicts(kl.pods)
 	case UPDATE:
-		glog.V(3).Infof("Update: Containers changed")
+		glog.Infof("Update: Containers changed")
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: I'm guessing all these log changes are while it is a WIP?

**Comment by vmarmol** [comment]
#5032 is also small enough and may be useful in this refactoring. That will probably get merged in tomorrow, it only has minor comments.

**Comment by yujuhong** [comment]
Sorry, somehow I missed this issue. I have not touched the pod<->container syncing code, so I will re-assign to @vmarmol, who has already commented on the PR (thanks!).

**Comment by vmarmol** [comment]
I was already reviewing the other one so figured I should take a look :)

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -1025,137 +1014,204 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
...
-	podStatus, err := kl.GetPodStatus(podFullName, uid)
+	err := kl.MakeDataDirs(uid)
 	if err != nil {
-		glog.Errorf("Unable to get pod with name %q and uid %q info with error(%v)", podFullName, uid, err)
+		return err
+	}
+	
+	podStatus, podInfraContainerID, found := kl.getPodInfraContainer(podFullName, uid, dockerContainers)
+	if found {
+		containersToKeep[podInfraContainerID] = nil
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: We don't need this case I believe

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -1025,137 +1014,204 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
...
+						record.Eventf(ref, "unhealthy", "Liveness Probe Failed %v - %v", containerID, container.Name)
+					}
+					glog.Infof("pod %q container %q is unhealthy (probe result: %v). Container will be killed and re-created.", podFullName, container.Name, live)
 				} else {
-					kl.readiness.set(dockerContainer.ID, false)
-				}
-				if live == probe.Success {
-					containersToKeep[containerID] = empty{}
-					continue
+					createPodInfraContainer = true
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Not really anything wrong, but: We don't necessarily need to do this. Today we do for a lot of cases we shouldn't and was a source of issues for the last bug in this code path. May be worth to change that as a separate PR after this.

File: pkg/kubelet/kubelet.go [hunk 6]
```diff
@@ -1025,137 +1014,204 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
...
 				}
-				ref, ok := kl.getRef(containerID)
-				if !ok {
-					glog.Warningf("No ref for pod '%v' - '%v'", containerID, container.Name)
-				} else {
-					record.Eventf(ref, "unhealthy", "Liveness Probe Failed %v - %v", containerID, container.Name)
+				// If RestartPolicy is Always or OnFailure we restart containers that were running before we
+				// killed them when restarting Infra Container.
+			} else {  // createPodInfraContainer == true
+				if pod.Spec.RestartPolicy.Never == nil {
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: We can probably do this logic in one place, we do it in quite a few places today. Once we know we need to restart the podContainer we add all containers to the restart queue and check them one by one to remove them if they say restart never.

**Comment by vmarmol** [comment]
Overall I like the direction and the changes. I would urge for smaller changes with a promise of faster review. That is almost certainly me being weary of change in this path though :)

I have a few more comments around simplifying some of the flow, but I'll hold until you let me know it is ready for more review.

File: pkg/kubelet/kubelet.go (hunk 6 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: I thought about that. The issue is that we would also need to check for successful runs for restart on failure. It is doable, but I thought that it'll be more complex. If you think otherwise I can rewrite it in such way. In addition, AFAICT, restart policy is per pod, not per container.

File: pkg/kubelet/kubelet.go (hunk 5 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: I believe @dchen1107 is working on this particular issue. I believe it'll be much easier after this refactoring.

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: Right. I'm still getting used to Go way of specifying private methods...

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: Yup - makes debugging much easier, and I couldn't find a way of passing -v=X to test-go.sh... I'll revert those when creating 'real' RP.

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: I think we do. Otherwise we'll delete infra container when deleting everything we don't want to keep.

**Comment by gmarek** [comment]
I managed to write new WIP version of this PR. As previously I didn't fix tests, but this time everything seems to be working (played with a cluster a little bit).

This one is big, because it splits the logic into analysis and application parts AND changes how we handle Infra Containers (they'll die if they're only thing that's alive). As the split already pretty much rewrites the whole logic there, I think there's no reason to do it in two separate steps.

PTAL. I find the logic there hard to follow (those 'continues'...) so I'm pretty sure I missed some corner case there.

Two cases to consider:
- differentiating starting fresh pod from the situation when Infra Pod died when RestartPolicy is Never, and we wan't to kill all containers ("shouldContainerBeRestarted" will return true, as it won't be able to find any past failures).
- case of 'paused' containers: I believe we completely ignore them right now.

**Comment by gmarek** [comment]
Friendly ping?

File: pkg/kubelet/kubelet.go [hunk 7]
```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
 	return containerID, nil
 }
 
-func (kl *Kubelet) syncPod(pod *api.BoundPod, containersInPod dockertools.DockerContainers) error {
+type podContainerChangesSpec struct {
+	StartInfraContainer bool
+	InfraContainerId    dockertools.DockerID
+	ContainersToStart   map[int]empty
+	ContainersToKeep    map[dockertools.DockerID]int
```
Comment (wojtek-t) [inline pkg/kubelet/kubelet.go]: What represents the "int" in this map?

**Comment by vmarmol** [comment]
Sorry, I missed this yesterday somehow. I'll review. Needs a rebase btw.

File: pkg/kubelet/kubelet.go (hunk 7 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: Index of container for pod.Spec.Containers. It was the pointer before but it doesn't work.

File: pkg/kubelet/kubelet.go [hunk 8]
```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
 	return containerID, nil
 }
 
-func (kl *Kubelet) syncPod(pod *api.BoundPod, containersInPod dockertools.DockerContainers) error {
+type podContainerChangesSpec struct {
+	StartInfraContainer bool
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: nit: these should be lowercase since the type itself is not publicly exposed

File: pkg/kubelet/kubelet.go [hunk 9]
```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
...
-		return err
+		glog.V(2).Infof("No Infra Container for %q found. All containers will be restarted.", podFullName)
+		createPodInfraContainer = true
 	}
 
-	for _, container := range pod.Spec.Containers {
+	for index, container := range pod.Spec.Containers {
 		expectedHash := dockertools.HashContainer(&container)
 		dockerContainerName := dockertools.BuildDockerName(uid, podFullName, &container)
+		// If we're creating infra containere everything will be killed anyway
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Should this comment be inside the if? Before !createPodInfraContainer

File: pkg/kubelet/kubelet.go [hunk 10]
```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
...
-				if result == probe.Success {
-					containersInPod.RemoveContainerWithID(containerID)
-					continue
+			if !createPodInfraContainer {
+				// look for changes in the container.
+				containerChanged := hash != 0 && hash != expectedHash
+				if !containerChanged {
+					result, err := kl.probeContainer(pod, podStatus, container, dockerContainer)
+					if err != nil {
+						// I don't get this logic...
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Add a TODO to look into this (feel free to place me). I think we're trying to be conservative in that we will only restart the task if the probe was run and it was not successful.

File: pkg/kubelet/kubelet.go [hunk 11]
```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
 	return containerID, nil
 }
 
-func (kl *Kubelet) syncPod(pod *api.BoundPod, containersInPod dockertools.DockerContainers) error {
+type podContainerChangesSpec struct {
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Lets document these types please :)

File: pkg/kubelet/kubelet.go [hunk 12]
```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
...
+	StartInfraContainer bool
+	InfraContainerId    dockertools.DockerID
+	ContainersToStart   map[int]empty
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: I feel kinda weird using the index. WDYT of having it be a pointer to a container instead?

File: pkg/kubelet/kubelet.go (hunk 7 above)
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Looks like this can be a map to empty. I don't see us use the index.

File: pkg/kubelet/kubelet.go [hunk 13]
```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
...
+	podFullName := GetPodFullName(pod)
+	uid := pod.UID
+	containerChanges, err := kl.computePodContainerChanges(pod, containersInPod)
+	glog.V(3).Infof("Got container changes for pod %q: %+v", podFullName, containerChanges)
+	if err != nil {
+		return err
+	}
+
+	if containerChanges.StartInfraContainer || (len(containerChanges.ContainersToKeep) == 0 && len(containerChanges.ContainersToStart) == 0) {
+		glog.V(4).Infof("Killing Infra Container for %q", podFullName)
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Lets log which of these two cases happened so we know why we're killing it.

File: pkg/kubelet/kubelet.go [hunk 14]
```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
...
+		// Killing phase: if we want to start new infra container, or nothing is running kill everything (including infra container)
+		if podInfraContainer, found, _ := containersInPod.FindPodContainer(podFullName, uid, dockertools.PodInfraContainerName); found {
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Maybe we can do this killing in the function below. It will simplify this function and that function. Doesn't have to be in this PR.

File: pkg/kubelet/kubelet.go [hunk 15]
```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
...
+		}
+	} else {
+		// Otherwise kill any containers in this pod which are not specified as ones to keep.
+		for id, container := range containersInPod {
+			_, keep := containerChanges.ContainersToKeep[id]
+			if !keep {
+				glog.V(3).Infof("Killing unwanted container in pod: %+v", container)
+				err = kl.killContainer(container)
+				if err != nil {
+					glog.Errorf("Error killing container: %v", err)
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Lets add the container ID on this mesage

File: pkg/kubelet/kubelet.go [hunk 16]
```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
...
+	if containerChanges.StartInfraContainer && (len(containerChanges.ContainersToStart) > 0) {
+		ref, err = api.GetReference(pod)
+		if err != nil {
+			glog.Errorf("Couldn't make a ref to pod %q: '%v'", podFullName, err)
+		}
+		glog.Infof("Creating pod infra container for %q", podFullName)
+		podInfraContainerID, err = kl.createPodInfraContainer(pod)
 		if err != nil {
-			glog.Errorf("Error killing container: %v", err)
+			glog.Errorf("Failed to introspect pod infra container: %v; Skipping pod %q", err, podFullName)
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: "Failed to create"?

File: pkg/kubelet/kubelet.go [hunk 17]
```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
...
+			kl.recorder.Eventf(ref, "failedMount",
+				"Unable to mount volumes for pod %q: %v", podFullName, err)
 		}
+		glog.Errorf("Unable to mount volumes for pod %q: %v; skipping pod", podFullName, err)
+		return err
+	}
+
+	// Start everything
+	for container := range containerChanges.ContainersToStart {
+		glog.V(4).Infof("Creating container %+v", container)
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Lets print the container instead of just the index please

**Comment by vmarmol** [comment]
Mainly minor comments, looks good overall. Let me know how the e2e does, I think you said they all passed.

File: pkg/kubelet/kubelet.go (hunk 12 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: That was my original idea. I run into the problem that I don't really understand pointer semantics in Go. What I found out is that when you iterate over a slice you get a copy of an element, not the reference to it, and it's being reused in each iteration (at least that's the kind of behaviour I was seeing). I may try getting the pointer directly by &vector[index], but then I'm kind of scared what will happen if runtime will decide that it'll resize the slice...

In other words: in C++ it'd have been a pointer from the start, in Go I'm a bit afraid

File: pkg/kubelet/kubelet.go (hunk 7 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: We do, when we find a container which is changed, and thus we restart infra container. In this case we move 'containersToKeep' to 'containersToStart', which, sadly, has to have different types (if we wan't to even pretend to be efficient-ish). containersToKeep are running Docker containers, while containersToStart are specs of containers to start. Yuck...

File: pkg/kubelet/kubelet.go (hunk 11 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: Done. I guess this is a good place to describe semantics, so it's a long comment.

File: pkg/kubelet/kubelet.go (hunk 8 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: Done.

File: pkg/kubelet/kubelet.go (hunk 9 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: Yup

File: pkg/kubelet/kubelet.go (hunk 10 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: Yes, but I think that it should strongly depend on error type. Some of them may be transient, but I guess some of them are not.

File: pkg/kubelet/kubelet.go (hunk 13 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: Done.

File: pkg/kubelet/kubelet.go (hunk 14 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: Yeah, I thought about adding 'containersToKill' to changes struct to make everything explicit. This would move all this logic to the analysis part where it belong, but the cost is some information redundancy. I don't feel strongly about any one of those solutions...

File: pkg/kubelet/kubelet.go (hunk 15 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: It's just an int... IIUC container will contain everything, as it's a DockerContainer.

File: pkg/kubelet/kubelet.go (hunk 16 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: Thanks.

File: pkg/kubelet/kubelet.go (hunk 17 above)
Comment (gmarek) [inline pkg/kubelet/kubelet.go]: Right. As I mentioned it was a pointer before...

**Comment by gmarek** [comment]
I run e2e on other PR. For this one they fail, as they fail on the head right now...

File: pkg/kubelet/kubelet.go (hunk 12 above)
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Ah, I see it is a flat slice. SGTM

File: pkg/kubelet/kubelet.go (hunk 14 above)
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: At least the killAllContainersInPod including the infra container should be overall simplifying without needing the extra map no? We can do that in a followup, unless I'm missing something.

File: pkg/kubelet/kubelet.go (hunk 15 above)
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Sorry I meant `container.ID` so we know what container we were trying to kill. The glog above it is a v=3 log so it may not show up.

**Comment by vmarmol** [comment]
With these changes please squash commits and I think we're ready. We're only waiting on e2e.

**Comment by gmarek** [comment]
After fixing e2e tests I have 25 green and 2 pending.

**Comment by vmarmol** [comment]
@gmarek sorry to be a pain, but can we remove the "WIP" from the commit message :)

**Comment by gmarek** [comment]
Sorry. I should have done it myself. Done.

**Comment by vmarmol** [comment]
Thanks for putting up with the nits! Merging.
